src/lib/semaphore.js

```diff
@@ -3,36 +3,50 @@ export default class Semaphore {
     // Max number of concurrent operations the semaphore will allow
```

Comment:
Are we using the semaphore anywhere currently?

Reply:
yes there is a race condition that needed to be protected